src/resources.js

```diff
@@ -507,13 +507,23 @@ export class Resources {
    */
   mutateWork_() {
     // Read all necessary data before mutates.
+    // The height changing depends largely on the target element's position
+    // in the active viewport. We consider the active viewport the part of the
+    // visible viewport below 10% from the top and above 25% from the bottom.
+    // This is basically the portion of the viewport where the reader is most
+    // likely focused right now. The main goal is to avoid drastic UI changes
+    // in that part of the content. The elements below the active viewport are
+    // freely resized. The elements above the viewport are resized and request
+    // scroll adjustment to avoid active viewport changing without user's
```

Comment:
It stores the scrollHeight before resize of these elements, then calculates it after resize and assumes that the difference is the needed scroll adjustment.